tensorflow_datasets/image/arc.py

```diff
@@ -0,0 +1,126 @@
+# coding=utf-8
+# Copyright 2019 The TensorFlow Datasets Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""ARC dataset."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from absl import logging
+import json
+import numpy as np
+import os
+import tensorflow as tf
+import tensorflow_datasets.public_api as tfds
+
+_CITATION = """
+@misc{chollet_françois_2019,
+  title     = {The Measure of Intelligence},
+  url       = {https://arxiv.org/abs/1911.01547},
+  journal   = {arXiv.org},
+  author    = {François Chollet},
+  year      = {2019},
+  month     = {Nov}
+}
+"""
+
+_DESCRIPTION = """
+ARC can be seen as a general artificial intelligence benchmark, as a program
+synthesis benchmark, or as a psychometric intelligence test. It is targeted at
+both humans and artificially intelligent systems that aim at emulating a
+human-like form of general fluid intelligence.
+"""
+
+
+_BASE_URL = "https://github.com/fchollet/ARC/"
+_COMMIT = "bd9e2c934c83d00251b7b4781ffc38cd167c885f"
+_DL_URL = "{}zipball/{}/".format(_BASE_URL, _COMMIT)
+_DL_RESOURCE = tfds.download.Resource(
+    url=_DL_URL,
+    extract_method=tfds.download.ExtractMethod.ZIP)
+_EXTRACT_SUBDIR = "fchollet-ARC-{}".format(_COMMIT[:7])
+
+
+class ARC(tfds.core.GeneratorBasedBuilder):
+  """The Abstraction and Reasoning Corpus (ARC)."""
+
+  VERSION = tfds.core.Version("1.0.0")
+
+  def _info(self):
+    return tfds.core.DatasetInfo(
+        builder=self,
+        # This is the description that will appear on the datasets page.
+        description=_DESCRIPTION,
+        # tfds.features.FeatureConnectors
+        features=tfds.features.FeaturesDict({
+            # Grids are of varying size
+            "task_id": tfds.features.Text(),
+            "train": tfds.features.Sequence({
+                "input": tfds.features.Sequence(tfds.features.Sequence(tf.int32)),
+                "output": tfds.features.Sequence(tfds.features.Sequence(tf.int32)),
+            }),
+            "test": tfds.features.Sequence({
+                "input": tfds.features.Sequence(tfds.features.Sequence(tf.int32)),
+                "output": tfds.features.Sequence(tfds.features.Sequence(tf.int32)),
+            })
+        }),
+        # If there's a common (input, target) tuple from the features,
+        # specify them here. They'll be used if as_supervised=True in
+        # builder.as_dataset.
+        supervised_keys=None,
```

Comment:
`('train', 'test')`

Reply:
It's a bit more complicated, I think: input would be all pair inputs + all train pair outputs, and test would be all test outputs. I'm not sure this can be done with TFDS, unless I restructure the dataset entirely.

Reply:
I see, current version looks good, no need to change it, thanks.

Reply:
I looked into this a bit. Currently, the structure of each task is:

```python
{
    "train": {
        "input": input_grids,
        "output": output_grids,
    },
    "test": {
        "input": input_grids,
        "output": output_grids,
    }
}
```

In order to make this dataset compatible with `as_supervised=True`, I would have to restructure it to something like this:

```python
{
    "input": {
        "train_input": train_input_grids,
        "train_output": train_output_grids,
        "test_input": test_input_grids,
    },
    "target": {
        "test_output": test_output_grids,
    }
}
```

It's doable, but a bit ugly IMHO, and it doesn't respect the original dataset's structure, plus I'm not sure it really adds much value for the user, since this dataset is so different from regular image tasks (in fact, I'm not even sure it really belongs in the `image` category, perhaps there should be another category for special datasets like this one, that don't really fit in any category, or a category for abstract images, or reasoning).

Reply:
Yes, it's better to keep the original dataset structure. Keeping the following is better
```
{
    "train": {
        "input": input_grids,
        "output": output_grids,
    },
    "test": {
        "input": input_grids,
        "output": output_grids,
    }
}
```